Tests/ParameterEncoderTests.swift

```diff
@@ -0,0 +1,451 @@
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        XCTAssertEqual(result.value, "?foo?=?bar?")
+    }
+
+    func testThatSlashesInKeysAndValuesAreNotPercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let parameters = ["foo": "/bar/baz/qux"]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        XCTAssertEqual(result.value, "foo=/bar/baz/qux")
+    }
+
+    func testThatSpacesInKeysAndValuesArePercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let parameters = [" foo ": " bar "]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        XCTAssertEqual(result.value, "%20foo%20=%20bar%20")
+    }
+
+    func testThatPlusesInKeysAndValuesArePercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let parameters = ["+foo+": "+bar+"]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        XCTAssertEqual(result.value, "%2Bfoo%2B=%2Bbar%2B")
+    }
+
+    func testThatPercentsInKeysAndValuesArePercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let parameters = ["percent%": "%25"]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        XCTAssertEqual(result.value, "percent%25=%2525")
+    }
+
+    func testThatNonLatinCharactersArePercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let parameters = [
+            "french": "français",
+            "japanese": "日本語",
+            "arabic": "العربية",
+            "emoji": "😃"
+        ]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        let expectedParameterValues = [
+            "arabic=%D8%A7%D9%84%D8%B9%D8%B1%D8%A8%D9%8A%D8%A9",
+            "japanese=%E6%97%A5%E6%9C%AC%E8%AA%9E",
+            "french=fran%C3%A7ais",
+            "emoji=%F0%9F%98%83"
+        ].joined(separator: "&")
+        XCTAssertEqual(result.value, expectedParameterValues)
+    }
+
+    func testStringWithThousandsOfChineseCharactersIsPercentEscaped() {
+        // Given
+        let encoder = URLEncodedFormEncoder()
+        let repeatedCount = 2_000
+        let parameters = ["chinese": String(repeating: "一二三四五六七八九十", count: repeatedCount)]
+
+        // When
+        let result = Result<String> { try encoder.encode(parameters) }
+
+        // Then
+        let escaped = String(repeating: "%E4%B8%80%E4%BA%8C%E4%B8%89%E5%9B%9B%E4%BA%94%E5%85%AD%E4%B8%83%E5%85%AB%E4%B9%9D%E5%8D%81",
+                             count: repeatedCount)
+        let expected = "chinese=\(escaped)"
+        XCTAssertEqual(result.value, expected)
+    }
```

Comment:
Looking good @jshier! With the complexity of all the new logic for url encoding, I wanted to make sure we have all the right tests. I enabled code coverage and found we're only at 43% on the `ParameterEncoder` file. This strikes me as complex enough that we should really shoot for 100% code coverage. Otherwise I'm afraid we'll end up with issues that could be very difficult to track down and will be easy to cause other regressions. Could you take a stab at bumping it up?

Reply:
I've been able to bump it up to 70%. Manually testing the various nested containers looks tricky.